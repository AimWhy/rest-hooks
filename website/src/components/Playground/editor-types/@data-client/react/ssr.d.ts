import * as _data_client_core from '@data-client/core';
import { Manager, State, Controller } from '@data-client/core';
import * as react_jsx_runtime from 'react/jsx-runtime';

declare function createPersistedStore(managers?: Manager[]): readonly [({ children }: {
    children: React.ReactNode;
}) => react_jsx_runtime.JSX.Element, () => State<unknown>, Controller<_data_client_core.DataClientDispatch>, undefined<State<unknown>, _data_client_core.ActionTypes, unknown> & {
    dispatch: unknown;
}]

declare const awaitInitialData: (id?: string) => Promise<any>;
declare const getInitialData: (id?: string) => any;

declare function createServerDataComponent(useReadyCacheState: () => State<unknown>, id?: string): ({ nonce }: {
    nonce?: string | undefined;
}) => react_jsx_runtime.JSX.Element;

declare const ServerData: ({ data, nonce, id, }: {
    data: State<unknown>;
    id?: string;
    nonce?: string | undefined;
}) => react_jsx_runtime.JSX.Element | null;

export { ServerData, awaitInitialData, createPersistedStore, createServerDataComponent, getInitialData };
