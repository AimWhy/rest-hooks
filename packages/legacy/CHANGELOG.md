# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## 1.0.0 (2019-12-02)


### ⚠ 💥 BREAKING CHANGES

* New packages should be ready for 1.0 release

### 🚀 Features

* New @rest-hooks/legacy package ([#187](https://github.com/coinbase/rest-hooks/issues/187)) ([78c9321](https://github.com/coinbase/rest-hooks/commit/78c9321f3560d2ea57b4c8478e9bdd789762ae13))


### 📝 Documentation

* Improve readme for new packages ([2406f2c](https://github.com/coinbase/rest-hooks/commit/2406f2c78a10e41f6aa1e7deeb4c957a3c94314d))



# Change Log

This project adheres to [Semantic Versioning](http://semver.org/).
Every release, along with the migration instructions, is documented on the Github [Releases](https://github.com/coinbase/rest-hooks/releases) page.
