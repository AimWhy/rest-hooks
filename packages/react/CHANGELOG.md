# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [7.0.0](https://github.com/coinbase/rest-hooks/compare/@rest-hooks/react@6.0.0...@rest-hooks/react@7.0.0) (2022-11-16)

### ⚠ 💥 BREAKING CHANGES

* Hooks may fetch more than previously

### 🚀 Features

* Revalidate on react-navigation focus for useFetch() ([6eba85a](https://github.com/coinbase/rest-hooks/commit/6eba85a80aef919b066650e3a7f6ff38af831f8b))

### [0.2.1](https://github.com/coinbase/rest-hooks/compare/@rest-hooks/react@0.2.0...@rest-hooks/react@0.2.1) (2022-11-13)

### 📝 Documentation

* Update badge links ([e79c685](https://github.com/coinbase/rest-hooks/commit/e79c6853e9414127c6eeaee784dc3f33546b9630))

## 0.2.0 (2022-11-13)

### 🚀 Features

* Add packages react and redux ([#2270](https://github.com/coinbase/rest-hooks/issues/2270)) ([6e28f5f](https://github.com/coinbase/rest-hooks/commit/6e28f5f465b6f4f5d444b56234f212863aeade31))
