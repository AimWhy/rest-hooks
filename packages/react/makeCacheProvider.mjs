export * from './dist/makeCacheProvider.js';
