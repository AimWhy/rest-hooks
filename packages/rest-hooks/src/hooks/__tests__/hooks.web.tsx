import { shapeToEndpoint } from '@rest-hooks/legacy';
import {
  StateContext,
  State,
  ActionTypes,
  actionTypes,
  __INTERNAL__,
  Controller,
  ControllerContext,
} from '@rest-hooks/react';
import { CacheProvider } from '@rest-hooks/react';
import { render } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import {
  CoolerArticleResource,
  PaginatedArticleResource,
  ArticleResourceWithOtherListUrl,
  StaticArticleResource,
} from '__tests__/legacy-3';
import nock from 'nock';
import React, { Suspense, useEffect } from 'react';
// relative imports to avoid circular dependency in tsconfig references

import { useRetrieve } from '..';
import { makeRenderRestHook, mockInitialState } from '../../../../test';
import { users, articlesPages, payload } from '../test-fixtures';

const { initialState } = __INTERNAL__;

const { INVALIDATE_TYPE, RESET_TYPE } = actionTypes;

async function testDispatchFetch(
  Component: React.FunctionComponent<any>,
  payloads: any[],
) {
  const dispatch = jest.fn();
  const tree = (
    <ControllerContext.Provider value={new Controller({ dispatch })}>
      <Suspense fallback={null}>
        <Component />
      </Suspense>
    </ControllerContext.Provider>
  );
  render(tree);
  expect(dispatch).toHaveBeenCalledTimes(payloads.length);
  let i = 0;
  for (const call of dispatch.mock.calls) {
    delete call[0]?.meta?.createdAt;
    delete call[0]?.meta?.promise;
    expect(call[0]).toMatchSnapshot();
    const action = call[0];
    const res = await action.payload();
    expect(res).toEqual(payloads[i]);
    i++;
  }
}

function testRestHook(
  callback: () => void,
  state: State<unknown>,
  dispatch = (v: ActionTypes) => Promise.resolve(),
) {
  return renderHook(callback, {
    wrapper: function Wrapper({ children }: { children: React.ReactNode }) {
      return (
        <ControllerContext.Provider value={new Controller({ dispatch })}>
          <StateContext.Provider value={state}>
            {children}
          </StateContext.Provider>
        </ControllerContext.Provider>
      );
    },
  });
}

let mynock: nock.Scope;

beforeAll(() => {
  nock(/.*/)
    .persist()
    .defaultReplyHeaders({
      'Access-Control-Allow-Origin': '*',
      'Content-Type': 'application/json',
    })
    .options(/.*/)
    .reply(200);
  mynock = nock(/.*/).defaultReplyHeaders({
    'Access-Control-Allow-Origin': '*',
    'Content-Type': 'application/json',
  });
});

afterAll(() => {
  nock.cleanAll();
});

describe('useRetrieve', () => {
  let renderRestHook: ReturnType<typeof makeRenderRestHook>;
  beforeEach(() => {
    mynock.get(`/article-cooler/${payload.id}`).reply(200, payload);
    mynock.get(`/article-static/${payload.id}`).reply(200, payload);
    mynock.get(`/user/`).reply(200, users);
    renderRestHook = makeRenderRestHook(CacheProvider);
  });
  afterEach(() => {
    nock.cleanAll();
  });

  it('should dispatch singles', async () => {
    function FetchTester() {
      useRetrieve(CoolerArticleResource.detailShape(), payload);
      return null;
    }
    await testDispatchFetch(FetchTester, [payload]);
  });

  it('should not dispatch will null params', () => {
    const dispatch = jest.fn();
    let params: any = null;
    const { rerender } = testRestHook(
      () => {
        useRetrieve(CoolerArticleResource.detailShape(), params);
      },
      initialState,
      dispatch,
    );
    expect(dispatch).toBeCalledTimes(0);
    params = payload;
    rerender();
    expect(dispatch).toBeCalled();
  });

  it('should dispatch with resource defined dataExpiryLength', async () => {
    function FetchTester() {
      useRetrieve(StaticArticleResource.detailShape(), payload);
      return null;
    }
    await testDispatchFetch(FetchTester, [payload]);
  });

  it('should dispatch with fetch shape defined dataExpiryLength', async () => {
    function FetchTester() {
      useRetrieve(StaticArticleResource.longLivingRequest(), payload);
      return null;
    }
    await testDispatchFetch(FetchTester, [payload]);
  });

  it('should dispatch with fetch shape defined errorExpiryLength', async () => {
    function FetchTester() {
      useRetrieve(StaticArticleResource.neverRetryOnErrorRequest(), payload);
      return null;
    }
    await testDispatchFetch(FetchTester, [payload]);
  });

  it('should not refetch after expiry and render', async () => {
    let time = 1000;
    global.Date.now = jest.fn(() => time);
    nock.cleanAll();
    const fetchMock = jest.fn(() => payload);
    mynock.get(`/article-cooler/${payload.id}`).reply(200, fetchMock).persist();
    const initialFixtures: any[] = [
      {
        endpoint: shapeToEndpoint(CoolerArticleResource.detailShape()),
        args: [payload],
        response: payload,
      },
    ];
    const { result, rerender } = renderRestHook(
      () => {
        return useRetrieve(CoolerArticleResource.detailShape(), payload);
      },
      { initialFixtures },
    );
    await result.current;
    expect(fetchMock).toHaveBeenCalledTimes(0);
    time += 100;
    rerender();
    await result.current;
    expect(fetchMock).toHaveBeenCalledTimes(0);
    // eslint-disable-next-line require-atomic-updates
    time += 610000000;
    rerender();
    await result.current;
    rerender();
    await result.current;
    expect(fetchMock).toHaveBeenCalledTimes(0);
  });
});
