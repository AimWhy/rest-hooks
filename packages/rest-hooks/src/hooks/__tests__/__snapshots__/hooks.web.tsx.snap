// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useRetrieve should dispatch singles 1`] = `
{
  "meta": {
    "args": [
      {
        "content": "whatever",
        "id": 5,
        "tags": [
          "a",
          "best",
          "react",
        ],
        "title": "hi ho",
      },
      undefined,
    ],
    "key": "GET http://test.com/article-cooler/5",
    "options": {
      "errorPolicy": [Function],
      "fetchInit": {},
      "getFetchInit": [Function],
      "method": "GET",
      "signal": undefined,
      "url": [Function],
      "useFetchInit": [Function],
    },
    "reject": [Function],
    "resolve": [Function],
    "schema": [Function],
    "throttle": true,
    "type": "read",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;

exports[`useRetrieve should dispatch with fetch shape defined dataExpiryLength 1`] = `
{
  "meta": {
    "args": [
      {
        "content": "whatever",
        "id": 5,
        "tags": [
          "a",
          "best",
          "react",
        ],
        "title": "hi ho",
      },
      undefined,
    ],
    "key": "GET http://test.com/article-static/5",
    "options": {
      "dataExpiryLength": 3600000,
      "errorPolicy": [Function],
      "fetchInit": {},
      "getFetchInit": [Function],
      "method": "GET",
      "signal": undefined,
      "url": [Function],
      "useFetchInit": [Function],
    },
    "reject": [Function],
    "resolve": [Function],
    "schema": [Function],
    "throttle": true,
    "type": "read",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;

exports[`useRetrieve should dispatch with fetch shape defined errorExpiryLength 1`] = `
{
  "meta": {
    "args": [
      {
        "content": "whatever",
        "id": 5,
        "tags": [
          "a",
          "best",
          "react",
        ],
        "title": "hi ho",
      },
      undefined,
    ],
    "key": "GET http://test.com/article-static/5",
    "options": {
      "errorExpiryLength": Infinity,
      "errorPolicy": [Function],
      "fetchInit": {},
      "getFetchInit": [Function],
      "method": "GET",
      "signal": undefined,
      "url": [Function],
      "useFetchInit": [Function],
    },
    "reject": [Function],
    "resolve": [Function],
    "schema": [Function],
    "throttle": true,
    "type": "read",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;

exports[`useRetrieve should dispatch with resource defined dataExpiryLength 1`] = `
{
  "meta": {
    "args": [
      {
        "content": "whatever",
        "id": 5,
        "tags": [
          "a",
          "best",
          "react",
        ],
        "title": "hi ho",
      },
      undefined,
    ],
    "key": "GET http://test.com/article-static/5",
    "options": {
      "errorPolicy": [Function],
      "fetchInit": {},
      "getFetchInit": [Function],
      "method": "GET",
      "signal": undefined,
      "url": [Function],
      "useFetchInit": [Function],
    },
    "reject": [Function],
    "resolve": [Function],
    "schema": [Function],
    "throttle": true,
    "type": "read",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;
